core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1705,1257 +2684,116 @@ object Stream {
-      * be interrupted at that point. Early termination of `that` does not terminate the resulting stream.
-      *
-      * Any errors that occur in either `this` or `that` stream result in the overall stream terminating
-      * with an error.
+      * Synchronously sends values through `sink`.
       *
-      * Upon finalization, the resulting stream will interrupt the background stream and wait for it to be
-      * finalized.
+      * If `sink` fails, then resulting stream will fail. If sink `halts` the evaluation will halt too.
       *
-      * This method is similar to `this mergeHaltL that.drain` but ensures the `that.drain` stream continues
-      * to be evaluated regardless of how `this` is evaluated or how the resulting stream is processed.
-      * This method is also similar to `Stream(this,that).join(2)` but terminates `that` upon termination of
-      * `this`.
+      * Note that observe will only output full segments of `O` that are known to be successfully processed
+      * by `sink`. So if Sink terminates/fail in midle of segment processing, the segment will not be available
+      * in resulting stream.
       *
       * @example {{{
-      * scala> import cats.effect.IO, scala.concurrent.ExecutionContext.Implicits.global
-      * scala> val data: Stream[IO,Int] = Stream.range(1, 10).covary[IO]
-      * scala> Stream.eval(async.signalOf[IO,Int](0)).flatMap(s => Stream(s).concurrently(data.evalMap(s.set))).flatMap(_.discrete).takeWhile(_ < 9, true).compile.last.unsafeRunSync
-      * res0: Option[Int] = Some(9)
+      * scala> import scala.concurrent.ExecutionContext.Implicits.global, cats.effect.IO, cats.implicits._
+      * scala> Stream(1, 2, 3).covary[IO].observe(Sink.showLinesStdOut).map(_ + 1).compile.toVector.unsafeRunSync
+      * res0: Vector[Int] = Vector(2, 3, 4)
       * }}}
       */
-    def concurrently[O2](that: Stream[F, O2])(implicit F: Concurrent[F]): Stream[F, O] =
-      Stream.eval(Deferred[F, Unit]).flatMap { interruptR =>
-        Stream.eval(Deferred[F, Option[Throwable]]).flatMap { doneR =>
-          Stream.eval(Deferred[F, Throwable]).flatMap { interruptL =>
-            def runR =
-              that
-                .interruptWhen(interruptR.get.attempt)
-                .compile
-                .drain
-                .attempt
-                .map { _.left.toOption }
-                .flatMap { r =>
-                  // to prevent deadlock, done must be signalled before `interruptL`
-                  // in case the interruptL is signaled before the `L` stream may be in
-                  // its `append` code, that requires `get` to complete, which won't ever complete,
-                  // b/c it will be evaluated after `interruptL`
-                  doneR.complete(r) >>
-                    r.fold(F.unit)(interruptL.complete)
+    def observe(sink: Sink[F, O])(implicit F: Concurrent[F]): Stream[F, O] =
```

Comment:
What's about `observe` that needs to be in InvariantOps? the use of `Sink`?

Also at the top of the file (can't comment there for some reason) it says:

> Most of the api is defined in InvariantOps

which can probably be removed

Reply:
I tried inlining this similar to `through` and `to` -- basically this:
```scala
def observe[F2[x] >: F[x]: Concurrent](sink: Stream[F, O] => Stream[F2, Unit]): Stream[F2, O] = ...
```

Unfortunately, this doesn't work because in the implementation of `observeAsync`, we do more than just call `sink(this)` -- rather, we create a `Queue[F2, ...]` and then call `sink(q.dequeue)`.